### PR TITLE
NMS-15776: Parallelize DNS lookups during audit phase

### PR DESCRIPTION
**Pull request by christianpape**
* JIRA: https://opennms.atlassian.net/browse/NMS-15776

These changes parallelize the DNS lookups for interface addresses in the audit phase. The process waits for the searches to complete in the final call to the completeModelImport() method. This means that timeouts do not add up and delay the entire syncronization process.

### CODE REVIEW

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/operations/SaveOrUpdateOperation.java [hunk 1]
```diff
@@ -124,6 +122,10 @@ public void foundInterface(InetAddress addr, Object descr, final PrimaryType pri
         //FIXME: verify this doesn't conflict with constructor.  The constructor already adds this
         //interface to the node.
         m_node.addIpInterface(m_currentInterface);
+
+        if (System.getProperty("org.opennms.provisiond.reverseResolveRequisitionIpInterfaceHostnames", "true").equalsIgnoreCase("true")) {
+            dnsLookups.add(CompletableFuture.supplyAsync(() -> getProvisionService().getHostnameResolver().getHostname(addr, m_node.getLocation().getLocationName())).thenAccept(s -> m_node.getInterfaceWithAddress(addr).setIpHostName(s)));
```
Comment (fooker) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/operations/SaveOrUpdateOperation.java:127]: Should we use some kind of thread pool here to limit this somehow?

Comment (christianpape) [reply opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/operations/SaveOrUpdateOperation.java:127]: Yes, I will work something in.

**Review by RangerRick — DISMISSED** [review]

**Review by fooker — APPROVED** [review]

File: docs/modules/operation/pages/deep-dive/provisioning/scalability.adoc [hunk 2]
```diff
@@ -88,3 +88,29 @@ For a large number of devices, you may want to set the `scan-interval` to `0` to
...
+[options="header"]
+[cols="1,3,2"]
+|===
+| Parameter
+| Description
+| Default value
+
+| org.opennms.netmgt.provision.dns.client.rpc.maxThreadPoolSize
+| Maximum thread pool size.
+|
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/provisioning/scalability.adoc:107]:
```suggestion
| 64
```

File: docs/modules/operation/pages/deep-dive/provisioning/scalability.adoc [hunk 3]
```diff
@@ -88,3 +88,29 @@ For a large number of devices, you may want to set the `scan-interval` to `0` to
...
+64
+| org.opennms.netmgt.provision.dns.client.rpc.queueCapacity
+| Maximum queue size for pending DNS lookups.
+|
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/provisioning/scalability.adoc:111]:
```suggestion
| 4096
```

File: docs/modules/operation/pages/deep-dive/provisioning/scalability.adoc [hunk 4]
```diff
@@ -88,3 +88,29 @@ For a large number of devices, you may want to set the `scan-interval` to `0` to
...
+4096
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/provisioning/scalability.adoc:112]:
```suggestion

```

File: docs/modules/operation/pages/deep-dive/provisioning/scalability.adoc [hunk 5]
```diff
@@ -88,3 +88,29 @@ For a large number of devices, you may want to set the `scan-interval` to `0` to
...
+|===
+
+NOTE: DNS reverse lookups can be disabled completely by setting the property `org.opennms.provisiond.reverseResolveRequisitionIpInterfaceHostnames` to `false`.
```
Comment (Bonrob2) [inline docs/modules/operation/pages/deep-dive/provisioning/scalability.adoc:115]:
```suggestion
NOTE: To completely disable DNS reverse lookups, set the property `org.opennms.provisiond.reverseResolveRequisitionIpInterfaceHostnames` to `false`.
```

**Review by Bonrob2 — COMMENTED** [review]
Minor text updates. Thank you!

**Review by fooker — APPROVED** [review]

**Review by Bonrob2 — APPROVED** [review]
Text LGTM! Thanks!

**Review by RangerRick — CHANGES_REQUESTED** [review]
the smoke test failures for this look legitimate, marking this as "request changes" just so it shows as a thing that needs doing before this can be merged

**Review by RangerRick — DISMISSED** [review]

**Comment by christianpape** [comment]
Converting to Draft, since the smoke-tests are still failing.

**Review by RangerRick — APPROVED** [review]

**Pull request merged**